src/Lecture2.hs

```diff
@@ -260,11 +437,33 @@ data EvalError
     = VariableNotFound String
     deriving (Show, Eq)
 
+maybeHead :: [a] -> Maybe a
+maybeHead [] = Nothing
+maybeHead (x:_) = Just x
+
+calculateAllLiterals :: Expr -> (Int, [String])
+calculateAllLiterals = go (0, [])
+  where go :: (Int, [String]) -> Expr -> (Int, [String])
+        go (value, variables) (Lit x) = (value + x, variables)
+        go (value, variables) (Var x) = (value, variables ++ [x])
+        go (value, variables) (Add augend addend) = (value + firstValue + secondValue
+                                                    , variables ++ firstVariables ++ secondVariables)
+          where (firstValue, firstVariables) = go (0, []) augend
+                (secondValue, secondVariables) = go (0, []) addend
```

Comment:
That's an interesting helper function that can be used both for `eval` and `constantFolding` 🤔 

`eval` can be implemented simpler without it but looks like an interesting approach 🙂